Displaying mode and controls to additional participants

mvbrock opened the pull request:
Per #44371, this PR modifies the session manager to send the mode and controls to the client joining the session. Both K8s and SSH sessions will inform users of the available controls:

```
% build/tsh --insecure kube join -m moderator f9d7154c-6300-4415-9519-fc8f3d0f14c3
Teleport > Creating session with ID: f9d7154c-6300-4415-9519-fc8f3d0f14c3...
Teleport > User matt2 joined the session with participant mode: peer.
Teleport > Waiting for required participants...
Teleport > Joining session with participant mode: moderator
Teleport > Controls
Teleport >   - CTRL-C: Leave the session
Teleport >   - t: Forcefully terminate the session
Teleport > Connecting to shell-pod over K8S
```

The session initiator will just see that the additional participant has joined the session:
```
Teleport > Creating session with ID: f9d7154c-6300-4415-9519-fc8f3d0f14c3...
Teleport > Waiting for required participants...
Teleport > User matt joined the session with participant mode: moderator.
Teleport > Connecting to shell-pod over K8S
root@shell-pod:/#
```

Changelog: A user joining a session will now see available controls for terminating & leaving the session
Fixes: https://github.com/gravitational/teleport/issues/44371

mvbrock committed:
Displaying mode and controls to additional participants

mvbrock committed:
Merge branch 'master' into mvbrock/display-controls-on-session-join

mvbrock commented:
@tigrato I closed the previous branch due to commit corruption. Let me know what you think about these changes, and if we should also apply them to the kube session manager.

mvbrock commented:
> <img alt="image" width="501" src="https://private-user-images.githubusercontent.com/7527103/366187089-d0648dcd-69dc-4199-b1e7-4250cc1f9b71.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU5OTc4OTUsIm5iZiI6MTcyNTk5NzU5NSwicGF0aCI6Ii83NTI3MTAzLzM2NjE4NzA4OS1kMDY0OGRjZC02OWRjLTQxOTktYjFlNy00MjUwY2MxZjliNzEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTBUMTk0NjM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTE5YTg5ZjA1ZjFiYjU3ZWMxMGY2Zjc5ZGVlMjkzMWJiZTg5ZTQwYzQ2ZjczZDk4NzUyMGZiYWUxYWI2ODE4MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Igxzto5u7DxJmlhBVtmIccG7Pvi1npZ4lkmn8_FQW4E">
> Why does the _terminate_ control show up in the 2nd example but not the first? Is this just a typo?

Whoops, yes, typo. Corrected.

mvbrock committed:
Merge branch 'master' into mvbrock/display-controls-on-session-join

tigrato commented on lib/srv/sess.go:
```diff
@@ -1278,7 +1279,6 @@ func (s *session) startInteractive(ctx context.Context, scx *ServerContext, p *p
 	s.io.AddReader("reader", inReader)
 	s.io.AddWriter(sessionRecorderID, utils.WriteCloserWithContext(scx.srv.Context(), s.Recorder()))
 	s.BroadcastMessage("Creating session with ID: %v", s.id)
-	s.BroadcastMessage(SessionControlsInfoBroadcast)
```
can we transform `SessionControlsInfoBroadcast`  into a template, move it out of srv and reuse for kube access?

mvbrock replied:
Definitely, that'll keep things clean.

mvbrock replied:
Well, with all the troubles I've had viewing kube sessions, I'll just move `SessionControlsInfoBroadcast` over to `kube/proxy/sess.go` for now. Let me know what you think @tigrato.

mvbrock replied:
After chatting w/ @tigrato I'm going to rework so SSH and K8s has the same experience.

mvbrock committed:
Moving SessionControlsInfoBroadcast over to kube/proxy

mvbrock committed:
Merge branch 'master' into mvbrock/display-controls-on-session-join

jakule reviewed (approved):
LGTM

tigrato commented on lib/kube/proxy/sess.go:
```diff
@@ -58,6 +58,10 @@ import (
 	"github.com/gravitational/teleport/lib/utils"
 )
 
+// SessionControlsInfoBroadcast is sent in tandem with session creation to inform any joining
+// users about the session controls.
+const SessionControlsInfoBroadcast = "Controls\r\n  - CTRL-C: Leave the session\r\n  - t: Forcefully terminate the session (moderators only)"
+
 const sessionRecorderID = "session-recorder"
```
Srv and kubernetes must share the same information. I am working with Matt to fix another issue he has

mvbrock replied:
Made changes to unify the SSH and K8s messages sent to the clients.

mvbrock committed:
Merge branch 'master' into mvbrock/display-controls-on-session-join

mvbrock committed:
Transitioning to consistent proxy-emitted mode+controls

mvbrock committed:
Moving message broadcast so new participant wont see it

mvbrock committed:
Merge branch 'master' into mvbrock/display-controls-on-session-join

mvbrock commented on lib/kube/proxy/forwarder.go:
```diff
@@ -1207,7 +1207,7 @@ func (f *Forwarder) join(ctx *authContext, w http.ResponseWriter, req *http.Requ
 		party := newParty(*ctx, stream.Mode, client)
 
-		err = session.join(party, true /* emitSessionJoinEvent */)
+		err = session.join(party)
```
Removed this flag since `newUser != initiatingUser` is equivalent and used in the `join(party)` implementation.

mvbrock replied:
Oh I see. How about emitting the audit event and controls message when `len(s.parties) > 0`, i.e. the session has already been initiated? This logic can apply to both the SSH and Kube session management, making them a little more similar. I added commit b66ad27 to show this, but can revert and just use the `emitSessionJoinEvent` flag if you think this is digging a bigger hole.

mvbrock replied:
Chatting with @tigrato, he mentioned that SSH sessions can stay alive with the initiator leaving and joining (whereas Kube sessions cannot), which makes the above logic problematic. I'll revert https://github.com/gravitational/teleport/commit/b66ad2727098b5e310f24b9b7aee7860959162e5 as mentioned above, and use the original `emitSessionJoinEvent` flag.

mvbrock commented on lib/kube/proxy/sess.go:
```diff
@@ -1017,10 +1018,17 @@ func (s *session) join(p *party, emitJoinEvent bool) error {
 	}
-
 	s.io.AddWriter(stringID, p.Client.stdoutStream())
-	s.BroadcastMessage("User %v joined the session with participant mode: %v.", p.Ctx.User.GetName(), p.Mode)
```
Moved so that the new participant doesn't see the redundant broadcast, e.g.:

```
Teleport > User matt joined the session with participant mode: moderator.
Teleport > Joining session with participant mode: moderator
```

mvbrock committed:
Possible unit test fix (cant seem to test locally)

zmb3 commented:
As a note for the future: some day I'd like to figure out a way to send this info "out of band" over some other channel. `tsh` would still end up displaying it in the session, but the web UI could put it somewhere other than inside the xterm window..

mvbrock committed:
Fixed unit test

mvbrock committed:
Merge branch 'master' into mvbrock/display-controls-on-session-join

mvbrock committed:
Adding a line break before messaging the participant

mvbrock committed:
Merge branch 'master' into mvbrock/display-controls-on-session-join

marcoandredinis committed:
Add User Tasks resource - protos (#46059)

* Add User Integration Tasks resource - protos

* add account id

* move state to task instead of instance

* rename from user integration task to user task

* add instance id

avatus committed:
Add notice to web UI that users arent equal to MAU (#46686)

This adds a dismissible notice to the Users page for usage based billing
users that notifies them that the user count here isn't an accurate
reflection of MAU

mvbrock committed:
Linter errors

ptgott committed:
Clarify TLS requirements in the Jira guide (#46484)

Closes #45654

- Indicate that certificates for the Jira web server cannot be self
  signed.
- Remove references to Caddy and a `Certificate` resource, which were
  left over from an attempted change to this guide that was not fully
  completed.

ravicious committed:
Remove TXT record validation of custom DNS zones in VNet (#46709)

* Remove TXT record validation from custom DNS zones

* Remove mentions of TXT records from docs

* Outline in the RFD why domain verification was dropped

* Update rfd/0163-vnet.md

Co-authored-by: Nic Klaassen <nic@goteleport.com>

---------

Co-authored-by: Nic Klaassen <nic@goteleport.com>

djohns7 committed:
docs: mention the --days flag when executing an audit log query (#45764)

* Update access-monitoring.mdx

Include the default date range in the CLI example. This range is otherwise unclear and is hidden in the tctl audit help menu.

* Update access-monitoring.mdx

* Update docs/pages/admin-guides/access-controls/access-monitoring.mdx

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

---------

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

ryanclark committed:
Remove access-graph path resolution, proxy `/enterprise` requests (#46541)

* Remove `access-graph` path from tsconfig

* Proxy /enterprise requested in dev

nklaassen committed:
update e ref (#46726)

nklaassen committed:
fix: tolerate mismatched key PEM headers (#46725)

* fix: tolerate mismatched key PEM headers

Issue #43381 introduced a regression where we now fail to parse PKCS8
encoded RSA private keys within an "RSA PRIVATE KEY" PEM block in
some cases.
This format is somewhat non-standard, usually PKCS8 data should be in a
"PRIVATE KEY" PEM block. However, certain versions of OpenSSL and
possibly even Teleport in specific cases have generated private keys in
this format.

This commit updates ParsePrivateKey and ParsePublicKey to be more
tolerant of PKCS8, PKCS1, or PKIX key data no matter which PEM header is
used.

Fixes #46710

changelog: fixed regression in private key parser to handle mismatched PEM headers

* fix typo in comment

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

---------

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

avatus committed:
Use dynamic base path for favicon images (#46719)

bernardjkim committed:
Add Datadog Incident Management plugin support (#46271)

* Implement datadog plugin

* Add unit tests

* Add fallback recipient config

* Rename to Datadog Incident Management

* Update tests

* Datadog Incident Management

* Update tctl resource plugin command

* Typos

* Lint

* Exclude api changes for now

* Set channel size

* Address feedback

- Add PluginShutdownTimeout const
- Support api endpoint configuration
- Add additional godocs/comments

* Comment about datadog client package

* Document Datadog API types

* Only post resolution message when the AR is resolved

* Fix lint

* Unused function

---------

Co-authored-by: hugoShaka <hugo.hervieux@goteleport.com>

vapopov committed:
Add AutoUpdate Client/Cache implementation (#46661)

* Add AutoUpdate Client/Cache implementation

* CR changes

* Add permission for proxy to access resources

* Rename all occurrences auto update to camelcase

* Remove auto update client wrapper

* Drop AutoUpdateServiceClient helper
Rename comments for consistency

marcoandredinis committed:
User Tasks: services and clients implementation (#46131)

This PR adds the implementation for the User Tasks:
- services (backend+cache)
- clients (API + tctl)
- light validation to set up the path for later PRs

eriktate committed:
expanding testplan for host user creation (#46729)

ptgott committed:
Fix operator docs reference generator bug (#46732)

In the reference page for one Kubernetes operator resource, some
Markdown links are malformed.

The issue is that some fields of custom resource definitions used by the
operator consist of arrays of anonymous objects with fields that are
also objects. When creating docs based on these fields, the operator
resource docs generator creates a malformed link reference.

This change modifies the generator to replace any spaces with hyphens
before outputting link references, causing the resulting internal links
to work correctly.

This change also does some light refactoring to remove an unnecessary
`switch` statement.

mvbrock committed:
Emitting audit event and controls message for additional parties, i.e. not the session initiator

teleport-post-release-automation[bot] committed:
[auto] Update AMI IDs for 16.4.0 (#46746)

Co-authored-by: GitHub <noreply@github.com>

strideynet committed:
Remove deprecated HTTP RemoteCluster endpoints (#46756)

* Remove deprecated HTTP RemoteCluster endpoints

* Remove redundant test

strideynet committed:
Add `tbot` helm chart to `version.mk` (#46763)

rosstimothy committed:
Remove LockConfiguration.LockName (#46772)

Cleans up the deprecated config option now that
https://github.com/gravitational/teleport.e/pull/5034 has been
merged.

eriktate committed:
adding a reference to  to the host user guide (#46765)

strideynet committed:
Replace more Logrus usage with Slog (#46757)

* Remove logrus from lib/auth/machineid

* Switch authclient.Config.Log and TunnelAuthDialerConfig.Log to Slog

* Add *slog.Logger to auth.Server

* Remove logrus usage in `lib/auth/access.go`

* Replace logrus with slog in lib/auth/accountrecovery.go

* Replace logrus with slog in `lib/auth/apiserver.go`

* Add missing logger to auth.Server

* Fix test

gabrielcorado committed:
Update AWS roles ARNs displayed on `tsh app login` for AWS console apps (#44983)

* feat(tsh): list aws console logins from server

* chore(services): remove unified resources change

This is being covered on another PR.

* test(tsh): solve TestAzure flakiness by waiting using app servers are ready

* fix(tsh): apps with logins were fallingback into using aws arns

* refactor(client): use GetEnrichedResources

* chore(client): rename function

* refactor(tsh): directly resource lisiting for apps and reuse cluster client

* chore(client): reset client changes

* refactor(tsh): reuse cluster client for fetching allowed logins

* chore(tsh): remove unused function param

* refactor(tsh): update getApp retry with login

* refactor(tsh): use a single function to grab profile and cluste client

* refactor(tsh): perform retry with login at caller site

* fix(tsh): close auth client

* test(tsh): fix test failing due to login misconfiguration

* test(tsh): fix lint errors

* test(tsh): remove unused imports

mvbrock committed:
Revert "Emitting audit event and controls message for additional parties, i.e. not the session initiator"

This reverts commit b66ad2727098b5e310f24b9b7aee7860959162e5.

fspmarshall committed:
bulk audit event export api (#46399)

mvbrock committed:
Reverting back to using the emitSessionJoin boolean

mvbrock committed:
Merge branch 'master' into mvbrock/display-controls-on-session-join

mvbrock committed:
Merge remote-tracking branch 'origin/master' into mvbrock/display-controls-on-session-join

mvbrock committed:
Merge branch 'master' into mvbrock/display-controls-on-session-join

tigrato commented on lib/kube/proxy/forwarder.go:
```diff
@@ -1207,7 +1207,7 @@ func (f *Forwarder) join(ctx *authContext, w http.ResponseWriter, req *http.Requ
 		client := &websocketClientStreams{uuid.New(), stream}
 		party := newParty(*ctx, stream.Mode, client)
 
-		err = session.join(party, true /* emitSessionJoinEvent */)
```
Please re-add it since not all ides show it

tigrato reviewed (approved):
Left one nit

mvbrock committed:
Nits and removing a debug log

mvbrock committed:
Merge branch 'master' into mvbrock/display-controls-on-session-join
